### DISCUSSION TITLE
Quantized input params from nntool

### DISCUSSION THREAD

**Discussion by finnfi** [bitcraze/discussions#146] [Q&A] [answered]
Hi! 

This may be a stupid question, but after I have quantized my model with the nntool, where can I check the quantization parameters I have to apply before I send my input to the neural network on the gap 8. 

If I understand it correctly, the images taken from the himax camera can be sent in directly since they already are between 0-255. But I do have a second input which are state values (floats), and I do believe I have to quantize them before I send them into the network. 

This may be a question for the GreenWaves people, but I need the answer pretty fast and this forum is more active. 

As always, I appreciate every help! 

Finn

**Comment by knmcguire** [comment]
Ah good question but not sure if I'll be able to help. It is indeed a question that greenwaves would be better to help but I'll give it a go!

As I believe that the nntool works, at least what you can see from [the classification example](https://github.com/bitcraze/aideck-gap8-examples/tree/master/examples/ai/classification), that it does not need to have the 'result' of the pictures used for quantization, only the pictures. You can see it [here in the nntool script](aquant samples/* --scheme SQ8). What it does, it that it first uses the full network to produce an output with these test picture, and that same output is used to quantize the network using those images as input.

I don't know if that makes sense? Because the original network already provides the output aka state, you won't need to give it again as it already trained on it.

**Reply by knmcguire** [reply]
But let me know if I understood your question wrongly

**Reply by finnfi** [reply]
Thanks for the fast reply. 
Yes I also understand it like that. 
I believe we have a misunderstanding on what we mean with "state". We try to implement a depth estimation network on the gap8, so we have 3 inputs: 2 pictures taken with the himax camera and a difference in estimated state of the drone at each picture from the kalman filter (received from the stm). The last input is what I mean by "state". We created a lot of training data, and managed to create a tf network which does an okay job on the validation set. Now I have to implement it on the gap8 (which is heavy since storing two pictures at once already eats a lot of memory!). I have managed to get the neural network up and running, but the output is trash. I believe it can be something wrong with how the input is fed. So for the pictures i believe we have to add an imageformatter to convert the values from [0:255] to [-128:127] which the AT kernels are working on. And the states have to be quantized somehow before they are fed into the network. I believe I need a multiplication factor which I have to apply to the state before I cast it to an int8_t. After some more research I believe I found it with help of the qshow command in the nntool.

**Reply by knmcguire** [reply]
Ah good, happy to see that you found the answer already! Could you perhaps suggest an answer here below with how you solved it in the end? I believe more people will be interested in it.

Interesting application btw! If you get further in your research, let us know if you would like to write an blogpost in the future.

**Reply by finnfi** [reply]
I answered below.
I will let you know!

**Reply by knmcguire** [reply]
Thanks!

**Comment by finnfi** [comment, marked as answer]
So I found the answer in an old Greenwaves gap_sdk issue: https://github.com/GreenWaves-Technologies/gap_sdk/issues/192#issuecomment-786061525. 
The quantization values can be found in the qshow.txt file when using the nntool!
